agent/engine/docker_task_engine.go

```diff
@@ -1238,12 +1240,137 @@ func (engine *DockerTaskEngine) GetDaemonManagers() map[string]dm.DaemonManager
 func (engine *DockerTaskEngine) pullContainerManifest(
 	task *apitask.Task, container *apicontainer.Container,
 ) dockerapi.DockerContainerMetadata {
-	// Currently a no-op
-	logger.Debug("Manifest pull is currently a no-op", logger.Fields{
-		field.TaskID:    task.GetID(),
-		field.Container: container.Name,
-		field.Image:     container.Image,
+	if container.IsInternal() {
+		// internal containers are not in-scope of digest resolution
+		return dockerapi.DockerContainerMetadata{}
+	}
+	// AppNet Agent container image is managed at start up so it is not in-scope for digest resolution.
+	// (it uses the same image as AppNet Relay container)
+	if task.IsServiceConnectEnabled() && container == task.GetServiceConnectContainer() {
+		return dockerapi.DockerContainerMetadata{}
+	}
+
+	var imageManifestDigest digest.Digest
+	digestFromPayload := referenceutil.GetDigestFromImageRef(container.Image)
+	if digestFromPayload != "" {
+		// Digest available in task payload
+		imageManifestDigest = digestFromPayload
+		logger.Info("Found image manifest digest in task payload", logger.Fields{
+			field.TaskARN:       task.Arn,
+			field.ContainerName: container.Name,
+			field.Image:         container.Image,
+			field.ImageDigest:   imageManifestDigest.String(),
+		})
+	} else if !engine.imagePullRequired(engine.cfg.ImagePullBehavior, container, task.GetID()) {
+		// Image pull is not required for the container so we will use a locally available
+		// image for the container. Get digest from a locally available image.
+		imgInspect, err := engine.client.InspectImage(container.Image)
+		if err != nil {
+			logger.Error("Failed to inspect image to find repo digest", logger.Fields{
+				field.TaskARN:       task.Arn,
+				field.ContainerName: container.Name,
+				field.Image:         container.Image,
+			})
+			return dockerapi.DockerContainerMetadata{
+				Error: dockerapi.CannotPullImageManifestError{FromError: err},
+			}
+		}
+		if len(imgInspect.RepoDigests) == 0 {
+			// Image was not pulled from a registry, so the user must have cached it on the
+			// host directly. Skip digest resolution for this case as there is no digest.
+			logger.Info("No repo digest found", logger.Fields{
+				field.TaskARN:       task.Arn,
+				field.ContainerName: container.Name,
+				field.Image:         container.Image,
+			})
+			return dockerapi.DockerContainerMetadata{}
+		}
+		parsedDigest := referenceutil.GetDigestFromImageRef(imgInspect.RepoDigests[0])
```

Comment:
Do we always want to get the first item in the `RepoDigests` list? What does it mean when there are multiple items in it?

Reply:
Updated.